### PR TITLE
S3CSI-28: Add File Permissions E2E Suite and Refactor Utilities for S3 CSI Driver Tests

### PR DESCRIPTION
**Pull request by anurag4DSB**
Note to reviewers: Easier when reviewing commit by commit. There is a story. 

This pull request introduces several enhancements and refactoring changes to the S3 CSI driver's end-to-end (E2E) test suite. The most notable changes include the addition of a new test suite for file permissions, the consolidation of constants for user/group IDs, and significant refactoring of utility functions to improve modularity and maintainability.

### Enhancements to Test Suites:
* **New File Permissions Test Suite**: Added a comprehensive test suite (`filepermissions.go`) to validate file permission behaviors, including default permissions, custom `file-mode` settings, permission inheritance, and edge-case behaviors like blocking `chmod` and `chown` operations.
* **Documentation Updates**: Updated the `README.md` for the E2E custom suites to include details about the new file permissions test suite and guidelines for running performance tests with the `--performance` flag. [[1]](diffhunk://#diff-65432201824ea55d7321f70ce43c46b5b55fc8b6ced0ec1b1c6ac8a510c747dbR15-R40) [[2]](diffhunk://#diff-65432201824ea55d7321f70ce43c46b5b55fc8b6ced0ec1b1c6ac8a510c747dbL107-R140)

### Code Refactoring:
* **Centralized Constants**: Moved `defaultNonRootUser` and `defaultNonRootGroup` constants to a shared location (`util.go`) and renamed them to `DefaultNonRootUser` and `DefaultNonRootGroup` for consistency. Updated all references across test files. [[1]](diffhunk://#diff-1ee3de2ee436062f617ec8ce33845664d9d27cf981f2e5bfdd0ce9d319b42834L200-R201) [[2]](diffhunk://#diff-bf02f123c1a343527e0d4b26024ec8b33a4975dba9d63c03aafb859ba20ab15cL22-L29) [[3]](diffhunk://#diff-bf02f123c1a343527e0d4b26024ec8b33a4975dba9d63c03aafb859ba20ab15cL116-R109) [[4]](diffhunk://#diff-bf02f123c1a343527e0d4b26024ec8b33a4975dba9d63c03aafb859ba20ab15cL140-R136) [[5]](diffhunk://#diff-8db30b394794ec959cad421776dd99e3d9ea70e7d939f542140ddeb390043486L92-R93)
* **Utility Function Enhancements**: Refactored utility functions in `util.go` to improve modularity. Introduced new helpers like `MaterialisePaths`, `CreateFileInPod`, and `CreateDirInPod` for batch operations and file/directory management within pods.

### Minor Improvements:
* **Imports Cleanup**: Added missing imports and organized existing ones in `util.go` for better readability and functionality.
* **Improved Logging and Error Handling**: Enhanced error messages and logging in utility functions like `createPod` and `createVolumeResourceWithMountOptions` for better debugging and maintainability.

These changes collectively improve the robustness, readability, and extensibility of the S3 CSI driver's E2E test suite.

### CODE REVIEW

File: tests/e2e/customsuites/filepermissions.go [hunk 1]
```diff
@@ -226,17 +157,16 @@ func (t *s3CSIFilePermissionsTestSuite) DefineTests(driver storageframework.Test
 		paths["file2"] = fmt.Sprintf("%s/file2.txt", paths["subdir2"])
 		paths["file3"] = fmt.Sprintf("%s/file3.txt", paths["subdir3"])
 
-		// Create directories
+		// Create directories using the helper function
```
Comment (dvasilas) [inline tests/e2e/customsuites/filepermissions.go:171]: nit: `using the helper function` is probably not adding value.

**Comment by dvasilas** [comment]
Last line of commit message of https://github.com/scality/mountpoint-s3-csi-driver/pull/35/commits/24f1d84a8eaf03d175e1a63b6d5b3e28651b4570 is `Remember blank line between title and body.` :laughing:

File: tests/e2e/customsuites/filepermissions.go [hunk 2]
```diff
@@ -0,0 +1,1401 @@
...
+	ginkgo.It("should not allow chmod to change file permissions (no-op)", func(ctx context.Context) {
+		res := createVolumeWithOptions(ctx, l.config, pattern, DefaultNonRootUser, DefaultNonRootGroup, "0600")
+		pod, err := CreatePodWithVolumeAndSecurity(ctx, f, res.Pvc, "", DefaultNonRootUser, DefaultNonRootGroup)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		defer e2epod.DeletePodWithWait(ctx, f.ClientSet, pod)
+
+		filePath := "/mnt/volume1/chmod-test-file"
+
+		// Create the test file
+		_, _, err = e2evolume.PodExec(f, pod, fmt.Sprintf(`sh -c 'echo aGVsbG8gd29ybGQ= | base64 -d > %s'`, filePath))
```
Comment (dvasilas) [inline tests/e2e/customsuites/filepermissions.go:862]: nit: You could use `CreateFileInPod`

File: tests/e2e/customsuites/filepermissions.go [hunk 3]
```diff
@@ -0,0 +1,1401 @@
...
+	ginkgo.It("should not allow chown to change file ownership (no-op)", func(ctx context.Context) {
+		res := createVolumeWithOptions(ctx, l.config, pattern, DefaultNonRootUser, DefaultNonRootGroup, "0600")
+		pod, err := CreatePodWithVolumeAndSecurity(ctx, f, res.Pvc, "", DefaultNonRootUser, DefaultNonRootGroup)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		defer e2epod.DeletePodWithWait(ctx, f.ClientSet, pod)
+
+		filePath := "/mnt/volume1/chown-test-file"
+
+		// Create the test file
+		_, _, err = e2evolume.PodExec(f, pod, fmt.Sprintf(`sh -c 'echo testcontent > %s'`, filePath))
```
Comment (dvasilas) [inline tests/e2e/customsuites/filepermissions.go:921]: nit: You could use `CreateFileInPod`.

**Review by dvasilas — APPROVED** [review]

File: tests/e2e/customsuites/util.go [hunk 4]
```diff
@@ -21,150 +23,207 @@ import (
...
+	for _, s := range specs {
+		sb.WriteString(fmt.Sprintf("mkdir -p %q\n", filepath.Dir(s.Path)))
+
+		if s.Content == "" && s.ContentB64 == "" {
+			// Directory
+			sb.WriteString(fmt.Sprintf("mkdir -p %q\n", s.Path))
+		} else {
+			b64 := s.ContentB64
+			if b64 == "" {
+				b64 = base64.StdEncoding.EncodeToString([]byte(s.Content))
```
[the remark this answers is not in the document]
Comment (fredmnl) [reply tests/e2e/customsuites/util.go:72]: Might be clearer to have a single Content field and a boolean flag that shows whether the content is b64encoded or not. Not sure which is best, your call.

**Review by fredmnl — APPROVED** [review]

File: tests/e2e/customsuites/util.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (anurag4DSB) [reply tests/e2e/customsuites/util.go:72]: Good catch! I've implemented your suggestion by adding an IsBinary flag to the PathSpec struct instead of the double encoding/decoding cycle. This makes the API clearer and more efficient - we now write text content directly using a heredoc and only use base64 encoding/decoding for binary data. I've also added a CreateBinaryFileInPod helper function to make working with binary files more convenient.

File: tests/e2e/customsuites/filepermissions.go [hunk 5]
```diff
@@ -0,0 +1,1401 @@
...
+	return t.tsInfo
+}
+
+// SkipUnsupportedTests is a no-op as all tests should be supported.
+func (t *s3CSIFilePermissionsTestSuite) SkipUnsupportedTests(_ storageframework.TestDriver, _ storageframework.TestPattern) {
+}
+
+// DefineTests implements the test suite functionality.
+func (t *s3CSIFilePermissionsTestSuite) DefineTests(driver storageframework.TestDriver, pattern storageframework.TestPattern) {
+	type local struct {
```
[the remark this answers is not in the document]
Comment (anurag4DSB) [reply tests/e2e/customsuites/filepermissions.go:68]: I made this change and became explicit in other methods as well.

File: tests/e2e/customsuites/filepermissions.go (hunk 2 above)
Comment (anurag4DSB) [reply tests/e2e/customsuites/filepermissions.go]: forgot this, thanks

File: tests/e2e/customsuites/filepermissions.go (hunk 3 above)
Comment (anurag4DSB) [reply tests/e2e/customsuites/filepermissions.go]: Thanks, great catch.

**Comment by anurag4DSB** [comment]
> Last line of commit message of [24f1d84](https://github.com/scality/mountpoint-s3-csi-driver/commit/24f1d84a8eaf03d175e1a63b6d5b3e28651b4570) is `Remember blank line between title and body.` 😆

OMG! I uncommented my git commit template. :D
